google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java

```diff
@@ -155,32 +162,70 @@ private void getOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOpt
          FirewallTagNamingConvention naming) {
 
       FirewallApi firewallApi = api.firewalls();
-      List<AtomicReference<Operation>> operations = Lists.newArrayList();
-
-      for (Integer port : templateOptions.getInboundPorts()) {
-         String name = naming.name(port);
-         Firewall firewall = firewallApi.get(name);
-         if (firewall == null) {
-            List<String> ports = ImmutableList.of(String.valueOf(port));
-            List<Rule> rules = ImmutableList.of(Rule.create("tcp", ports), Rule.create("udp", ports));
-            FirewallOptions firewallOptions = new FirewallOptions().name(name).network(network.selfLink())
-                  .allowedRules(rules).sourceTags(templateOptions.getTags())
-                  .sourceRanges(of(DEFAULT_INTERNAL_NETWORK_RANGE, EXTERIOR_RANGE)).targetTags(ImmutableList.of(name));
-            AtomicReference<Operation> operation = Atomics.newReference(firewallApi.createInNetwork(
-                  firewallOptions.name(), network.selfLink(), firewallOptions));
-            operations.add(operation);
-         }
+      int[] inboundPorts = templateOptions.getInboundPorts();
+      if ((inboundPorts == null) || inboundPorts.length == 0){
+         return;
       }
 
-      for (AtomicReference<Operation> operation : operations) {
+      List<String> ports = simplifyPorts(inboundPorts);
+      String name = naming.name(ports);
+      Firewall firewall = firewallApi.get(name);
+      AtomicReference<Operation> operation = null;
+      if (firewall == null){
```

Comment:
If firewall is not null then the specific firewall we are about to create already exists. We are appending a hash to the end of the firewall name to both make the name unique and indicate which ports it opens to tcp and udp. If a firewall with the name already exists then we assume it opens the correct ports and has the correct tag.